### PR TITLE
Implement Windows Unicode APIs

### PR DESCRIPTION
**Pull request by fireice-uk**
Background (or let's go back to 1993):

Windows internally uses USC2 encoding. This is a grandfather of UTF-16, back from the days where 16 bits was enough for everyone.

Every single windows API function has two variants. W and A - the latter calling the W function and **converting the result to local codepage**. This is important as this isn't a lossless conversion. If the USC2 glyphs are not in user's codepage, they cannot be recreated (as Monero guys attempted to do here https://github.com/ryo-currency/ryo-currency/pull/102/commits/5762596a4876e23a4e09d218b509692346d976ed#diff-cb82e0aa6db499004305bf35f0eaafbcL161)

So the key to unicode support on Windows is to make sure either we or the underlying library calls the W variant function and then converts to UTF8. Since Windows is best at handling Windows mess, we will use Microsoft's own conversion via `WideCharToMultiByte`.

This is a generic and recurring problem. Any string going in or out of the Windows API needs to be treated this way - command line args, console input, filenames etc.

### CODE REVIEW

File: src/simplewallet/simplewallet.cpp [hunk 1]
```diff
@@ -157,39 +158,37 @@ const command_line::arg_descriptor<bool> arg_use_english_language_names = {"use-
...
+
+	ReadConsoleW(hConIn, buffer, sizeof(buffer)/sizeof(wchar_t)-1, &read, nullptr);
+	buffer[read] = 0;
+
+	SetConsoleMode(hConIn, oldMode);
+	CloseHandle(hConIn);
+ 
+	int size_needed = WideCharToMultiByte(CP_UTF8, 0, buffer, -1, NULL, 0, NULL, NULL);
+	std::string buf(size_needed, '\0');
+	WideCharToMultiByte(CP_UTF8, 0, buffer, -1, &str[0], size_needed, NULL, NULL);
```
Comment (psychocrypt) [inline src/simplewallet/simplewallet.cpp]: maybe I am blind but where is `str` defined?

**Comment by fireice-uk** [comment]
Ready for testing, review and merge.

File: src/common/command_line.cpp [hunk 2]
```diff
@@ -90,6 +95,33 @@ bool is_no(const std::string &str)
...
+	{
+		int size_needed = WideCharToMultiByte(CP_UTF8, 0, szArgs[i], -1, NULL, 0, NULL, NULL);
+		args.emplace_back(size_needed, '\0');
+		std::string& str = args.back();
+		char* strptr = &str[0];
+		WideCharToMultiByte(CP_UTF8, 0, szArgs[i], -1, strptr, size_needed, NULL, NULL);
+		str.pop_back();
+		argptrs.emplace_back(strptr);
+	}
+}
```
Comment (psychocrypt) [inline src/common/command_line.cpp]: missing `return true;`

File: src/common/command_line.cpp [hunk 3]
```diff
@@ -90,6 +95,33 @@ bool is_no(const std::string &str)
...
+#ifdef WIN32
+bool get_windows_args(std::vector<std::string>& args, std::vector<char*>& argptrs)
+{
+	int nArgs = 0;
+	LPWSTR* szArgs = CommandLineToArgvW(GetCommandLineW(), &nArgs);
+
+	if(szArgs == nullptr)
+		return false;
+
+	for(int i=0; i < nArgs; i++)
```
Comment (psychocrypt) [inline src/common/command_line.cpp]: because you need to touch this file anyway could you please add spaces around the assign operator `int i = 0; ...`

**Pull request merged**